### PR TITLE
[WIP] Add Rails 5.2 & Capybara 3 support.

### PR DESCRIPTION
**Pull request by gobijan**
## Summary

Relax the runtime dependencies to allow railties version 5.2.

## Details

Relax the runtime dependencies to allow railties version 5.2.

## Motivation and Context

Allow Rails 5.2 now that it's released.

## How Has This Been Tested?

No introduction of new tests. Relying on tests that are already there.

## Screenshots (if appropriate):
none.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

- [ ] I've added tests for my code.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by mvz** [comment]
Maybe @gobijan and @roberts1000 can cooperate on this :wink:.

**Comment by gobijan** [comment]
Sure. I am experimenting on my PR and trying latest gem versions. @roberts1000 you have any findings to share for now?

File: cucumber-rails.gemspec [hunk 1]
```diff
@@ -12,23 +12,23 @@ Gem::Specification.new do |s|
...
-  s.add_development_dependency('aruba', '~> 0.14.2')
+  s.add_development_dependency('aruba', '>= 0.14.2')
   s.add_development_dependency('builder', ['>= 3.1.0', '< 4'])
   s.add_development_dependency('bundler', '>= 1.3.5')
   s.add_development_dependency('selenium-webdriver', '>= 3.4.1')
   s.add_development_dependency('database_cleaner', '>= 1.0.0')
   s.add_development_dependency('factory_girl', '>= 3.2')
   s.add_development_dependency('rake', '>= 0.9.2.2')
-  s.add_development_dependency('rspec', '~> 3.0')
+  s.add_development_dependency('rspec', '~> 3')
```
Comment (mvz) [inline cucumber-rails.gemspec]: This will allow RSpec 4, which is probably not what you want.

File: cucumber-rails.gemspec [hunk 2]
```diff
@@ -12,23 +12,23 @@ Gem::Specification.new do |s|
...
-  s.add_runtime_dependency('railties', ['>= 4', '< 5.2'])
+  s.add_runtime_dependency('nokogiri', '~> 1.8')
+  s.add_runtime_dependency('railties', ['>= 4', '< 6'])
   s.add_runtime_dependency('mime-types', ['>= 1.17', '< 4'])
 
   # Main development dependencies
   s.add_development_dependency('ammeter', ['>= 1.0.0', '!= 1.1.3'])
   s.add_development_dependency('appraisal', '>= 0.5.1')
-  s.add_development_dependency('aruba', '~> 0.14.2')
+  s.add_development_dependency('aruba', '>= 0.14.2')
```
Comment (mvz) [inline cucumber-rails.gemspec]: The latest release of aruba is in the 0.14.x series, so this change has no effect currently.

File: cucumber-rails.gemspec [hunk 3]
```diff
@@ -12,23 +12,23 @@ Gem::Specification.new do |s|
 
   s.license = 'MIT'
 
-  s.add_runtime_dependency('capybara', ['>= 1.1.2', '< 3'])
+  s.add_runtime_dependency('capybara', ['>= 1.1.2', '< 4'])
```
Comment (xtrasimplicity) [inline cucumber-rails.gemspec]: This also resolves @Kosmas' request here: https://github.com/cucumber/cucumber-rails/issues/375#issuecomment-380043499 :+1:

Comment (Kosmas) [reply cucumber-rails.gemspec]: :+1:

**Comment by roberts1000** [comment]
@gobijan Glad to see someone else is taking a look at this one!  I didn't hit the same test failures that you hit.  I tried to avoid changing gem/ruby versions if they weren't related to Rails 5.2 so that may the reason for the difference.  My PR is failing on a bootsnap loading error.

File: .travis.yml [hunk 4]
```diff
@@ -1,8 +1,7 @@
 rvm:
-  - 2.2.9
```
Comment (radar) [inline .travis.yml]: I would suggest keeping this, as cucumber-rails still is using Ruby 2.2. If Rails 5.2 requires higher, then I suggest excluding it below in `.travis.yml`.

File: cucumber-rails.gemspec [hunk 5]
```diff
@@ -12,23 +12,23 @@ Gem::Specification.new do |s|
...
   # Main development dependencies
   s.add_development_dependency('ammeter', ['>= 1.0.0', '!= 1.1.3'])
   s.add_development_dependency('appraisal', '>= 0.5.1')
-  s.add_development_dependency('aruba', '~> 0.14.2')
+  s.add_development_dependency('aruba', '>= 0.14.2')
   s.add_development_dependency('builder', ['>= 3.1.0', '< 4'])
   s.add_development_dependency('bundler', '>= 1.3.5')
   s.add_development_dependency('selenium-webdriver', '>= 3.4.1')
   s.add_development_dependency('database_cleaner', '>= 1.0.0')
   s.add_development_dependency('factory_girl', '>= 3.2')
```
Comment (radar) [inline cucumber-rails.gemspec]: Consider changing this to `factory_bot` while you're here too. There will need to be some other changes made to the features (like `emulate_javascript.feature` too)

**Comment by radar** [comment]
Keen to get this merged so that we can upgrade to Rails 5.2 as well. To help with the effort, I'm opening a few PRs on @gobijan's repo too:

https://github.com/gobijan/cucumber-rails/pull/1
https://github.com/gobijan/cucumber-rails/pull/2

File: cucumber-rails.gemspec (hunk 5 above)
Comment (radar) [reply cucumber-rails.gemspec]: I've just submitted a PR to your repo for this: https://github.com/gobijan/cucumber-rails/pull/2

**Comment by alaxalves** [comment]
Do yall have a forecast on dropping this??

**Comment by mvz** [comment]
@alaxalves there are still some tests failing in this pull request, so it's hard to tell.

**Comment by gobijan** [comment]
Puh. Feels a little stuck here. Some tests are failing. Anyone very familiar with the code base can have a look on what's going on? @mvz what do you think? Looking at the tests the 4.2 Gemfile is okay but the newer versions are currently failing.

**Comment by xtrasimplicity** [comment]
@gobijan Some of the failing tests are caused by Mozilla Geckodriver, which we use to handle communication between Selenium and Firefox.This is an unrelated issue (https://github.com/cucumber/cucumber-rails/issues/370), which can be _temporarily_ suppressed by restarting each individual failing job.

I should have some time this evening (AEST) to jump in and take a look at this PR to help get things moving a bit quicker. I've spent a bucket load of time working with this test suite in the past, so I'm hoping that its something that I've come across before so we can get this fixed up and rolled out asap.

**Comment by radar** [comment]
Thank you @xtrasimplicity for your efforts :)

**Comment by xtrasimplicity** [comment]
It appears that the Gemfile used in the Rails 5 app does not match the standard Rails 5 Gemfile. I made some minor temporary changes to the `rails_new` method to overwrite the `config/boot.rb` file and dump the contents of the `Gemfile` and the folder structure to stdout, as follows:
```ruby
def rails_new(options = {})
    options[:name] ||= 'test_app'
    command = run "bundle exec rails new #{options[:name]} --skip-bundle --skip-test-unit --skip-spring #{options[:args]}"
    expect(command).to have_output /README/
    expect(last_command_started).to be_successfully_executed
    cd options[:name]
    delete_environment_variable 'RUBYOPT'
    delete_environment_variable 'BUNDLE_BIN_PATH'
    delete_environment_variable 'BUNDLE_GEMFILE'

    overwrite_file File.join('config', 'boot.rb'), %Q{
      ENV['BUNDLE_GEMFILE'] ||= File.expand_path('../Gemfile', __dir__)
      begin
        require 'bundler/setup' # Set up gems listed in the Gemfile.
        require 'bootsnap/setup' # Speed up boot time by caching expensive operations.
      rescue LoadError
        puts `cat $BUNDLE_GEMFILE`
        puts `pwd`
        exit 1
      end
    }
  [...]
end
```
With this, I can see that the gemfile being used is:
```
source 'http://rubygems.org'
gem 'rails'
gem "cucumber-rails", group: :test, path: "../../.."
gem "capybara", group: :test
gem "rspec-rails", group: :test
if RUBY_VERSION >= '2.0.0'
  gem 'sass-rails'
  gem 'uglifier'
end
```
where a typical rails 5 app uses:
```
source 'https://rubygems.org'
git_source(:github) { |repo| "https://github.com/#{repo}.git" }

[...]

# Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
gem 'rails', '~> 5.2.0'
# Use sqlite3 as the database for Active Record
gem 'sqlite3'
# Use Puma as the app server
gem 'puma', '~> 3.11'
# Use SCSS for stylesheets
gem 'sass-rails', '~> 5.0'
# Use Uglifier as compressor for JavaScript assets
gem 'uglifier', '>= 1.3.0'
# See https://github.com/rails/execjs#readme for more supported runtimes
# gem 'mini_racer', platforms: :ruby

# Use CoffeeScript for .coffee assets and views
gem 'coffee-rails', '~> 4.2'
# Turbolinks makes navigating your web application faster. Read more: https://github.com/turbolinks/turbolinks
gem 'turbolinks', '~> 5'
# Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
gem 'jbuilder', '~> 2.5'
# Use Redis adapter to run Action Cable in production
# gem 'redis', '~> 4.0'
# Use ActiveModel has_secure_password
# gem 'bcrypt', '~> 3.1.7'

# Use ActiveStorage variant
# gem 'mini_magick', '~> 4.8'

# Use Capistrano for deployment
# gem 'capistrano-rails', group: :development

# Reduces boot times through caching; required in config/boot.rb
gem 'bootsnap', '>= 1.1.0', require: false

group :development, :test do
  # Call 'byebug' anywhere in the code to stop execution and get a debugger console
  gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
end

group :development do
  # Access an interactive console on exception pages or by calling 'console' anywhere in the code.
  gem 'web-console', '>= 3.3.0'
  gem 'listen', '>= 3.0.5', '< 3.2'
  # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
  gem 'spring'
  gem 'spring-watcher-listen', '~> 2.0.0'
end

group :test do
  # Adds support for Capybara system testing and selenium driver
  gem 'capybara', '>= 2.15', '< 4.0'
  gem 'selenium-webdriver'
  # Easy installation and use of chromedriver to run system tests with Chrome
  gem 'chromedriver-helper'
end

# Windows does not include zoneinfo files, so bundle the tzinfo-data gem
gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
```

**Comment by xtrasimplicity** [comment]
The issue was caused by the `No database` feature overwriting the Gemfile with a custom gemfile, in order to remove the 'sqlite' and 'database_cleaner' gems. Rather than having Aruba replacing the whole file, I've added a step to remove specific gems from the temporary application's `Gemfile`. This appears to have resolved the issue. :)
https://github.com/gobijan/cucumber-rails/pull/6

**Comment by xtrasimplicity** [comment]
Apart from the `geckodriver -v` issues in a few runs, it is looking pretty
good! Great work, everyone!

File: cucumber-rails.gemspec [hunk 6]
```diff
@@ -37,7 +37,7 @@ Gem::Specification.new do |s|
   s.add_development_dependency('rdoc', '>= 3.4')
   s.add_development_dependency('yard', '>= 0.8.7')
 
-  s.required_ruby_version = '>= 2.2.0'
+  s.required_ruby_version = '>= 2.3.0'
```
Comment (xtrasimplicity) [inline cucumber-rails.gemspec]: `Cucumber-ruby` supports `>= 2.2`, so we should change this back to keep things standardised.

File: features/emulate_javascript.feature [hunk 7]
```diff
@@ -16,12 +16,12 @@ Feature: Emulate Javascript
     And I write to "features/step_definitions/s.rb" with:
       """
       Given /^there is a widget named "([^"]*)"$/ do |name|
-        FactoryGirl.create(:widget, name: name)
+        FactoryBot.create(:widget, :name => name)
```
Comment (xtrasimplicity) [inline features/emulate_javascript.feature]: We shouldn't regress with this. `:key => value` is the old Ruby style. :)

File: features/emulate_javascript.feature [hunk 8]
```diff
@@ -73,12 +73,12 @@ Feature: Emulate Javascript
...
```
Comment (xtrasimplicity) [inline features/emulate_javascript.feature]: We shouldn't regress with this. `:key => value` is the old Ruby style. :)

File: cucumber-rails.gemspec [hunk 9]
```diff
@@ -26,9 +26,9 @@ Gem::Specification.new do |s|
   s.add_development_dependency('bundler', '>= 1.3.5')
   s.add_development_dependency('selenium-webdriver', '>= 3.4.1')
   s.add_development_dependency('database_cleaner', '>= 1.0.0')
-  s.add_development_dependency('factory_girl', '>= 3.2')
+  s.add_development_dependency('factory_bot', '>= 3.2')
   s.add_development_dependency('rake', '>= 0.9.2.2')
-  s.add_development_dependency('rspec', '~> 3.0')
+  s.add_development_dependency('rspec', '~> 3')
```
Comment (mvz) [inline cucumber-rails.gemspec]: This should become `~> 3.0` again to avoid allowing rspec 4 automatically when it arrives.

**Comment by mvz** [comment]
Almost there \o/

File: cucumber-rails.gemspec (hunk 9 above)
Comment (xtrasimplicity) [reply cucumber-rails.gemspec]: Woops. I'll fix that now - thanks @mvz!

**Comment by aslakhellesoy** [comment]
Hi @gobijan,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber

**Comment by xtrasimplicity** [comment]
All merged! The last change wouldn't have any impact on the test suite. :)

**Comment by gobijan** [comment]
Nice! Thanks everyone for contributing on this 😀 Was fun to see open source at it's best. 
Also thanks for adding me to the cucumber org @aslakhellesoy I'm happy to help out here and there :)

**Comment by radar** [comment]
Wonderful :)

I've opened a PR on https://github.com/DevicoSolutions/geckodriver-helper/pull/5 which should hopefully put an end to the intermittent builds too.

**Comment by radar** [comment]
In the meantime, could we please get a new gem release containing this PR's fixes? I believe that would allow codebases that use cucumber-rails to also upgrade to Rails 5.2.

**Comment by xtrasimplicity** [comment]
@radar, we'll need to finalise the changelog first (https://github.com/cucumber/cucumber-rails/pull/385), but once we've done this, we'll have to get @mvz, @olleolleolle, or @aslak to proceed with a release (when they're ready).

**Comment by mvz** [comment]
@xtrasimplicity That's a different Aslak :slightly_smiling_face:

**Comment by xtrasimplicity** [comment]
Woops! Thanks @mvz.

**Pull request merged**